### PR TITLE
FINERACT-2494: Add unit tests for ApiParameterHelper in fineract-core

### PR DESCRIPTION
**Pull request by Ambika-Sony**
## Description

This PR adds unit test coverage for the ApiParameterHelper utility class in the fineract-core module.

Changes:

Added ApiParameterHelperTest.java to test core methods like extractFieldsForResponseIfProvided.

Verified code formatting using ./gradlew spotlessApply.

Testing:

Successfully ran locally using: ./gradlew :fineract-core:test --tests "org.apache.fineract.infrastructure.core.api.ApiParameterHelperTest"

All tests passed (100% success rate).
## Checklist

Please make sure these boxes are checked before submitting your pull request - thanks!

- [ x] Write the commit message as per [our guidelines](https://github.com/apache/fineract/blob/develop/CONTRIBUTING.md#pull-requests)
- [ x] Acknowledge that we will not review PRs that are not passing the build _("green")_ - it is your responsibility to get a proposed PR to pass the build, not primarily the project's maintainers.
- [ x] Create/update [unit or integration tests](https://fineract.apache.org/docs/current/#_testing) for verifying the changes made.
- [ x] Follow our [coding conventions](https://cwiki.apache.org/confluence/display/FINERACT/Coding+Conventions).
- [ ] Add required Swagger annotation and update API documentation at fineract-provider/src/main/resources/static/legacy-docs/apiLive.htm with details of any API changes
- [ x] [This PR must not be a "code dump"](https://cwiki.apache.org/confluence/display/FINERACT/Pull+Request+Size+Limit). Large changes can be made in a branch, with assistance. Ask for help on the [developer mailing list](https://fineract.apache.org/#contribute).

Your assigned reviewer(s) will follow our [guidelines for code reviews](https://cwiki.apache.org/confluence/display/FINERACT/Code+Review+Guide).

### CODE REVIEW

**Comment by adamsaghy** [comment]
@Ambika-Sony Testing 1 method does not provide too much value. Would you consider to test the whole class?

**Comment by Ambika-Sony** [comment]
> @Ambika-Sony Testing 1 method does not provide too much value. Would you consider to test the whole class?

I've Implemented pagination (limit, offset), sorting (orderBy, sortOrder), and association helpers in ApiParameterHelper.

Added 10 unit tests in ApiParameterHelperTest, covering standard scenarios, null/empty inputs, and defensive checks for whitespace handling.

Verified the build locally with ./gradlew :fineract-core:test --tests ApiParameterHelperTest (all 10 passed).

Looking forward to your feedback.

**Comment by Ambika-Sony** [comment]
@adamsaghy I noticed the Cucumber E2E tests are failing in the CI, but since my changes are strictly limited to the ApiParameterHelper utility and its unit tests (2 files), these failures seem unrelated to my code. I saw your recent email about the E2E framework transition—perhaps these are known issues?

File: fineract-core/src/test/java/org/apache/fineract/infrastructure/core/api/ApiParameterHelperTest.java [hunk 1]
```diff
@@ -0,0 +1,131 @@
+/**
+*Licensed to the Apache Software Foundation (ASF) under one
```
Comment (Aman-Mittal) [inline fineract-core/src/test/java/org/apache/fineract/infrastructure/core/api/ApiParameterHelperTest.java]: Please fix the header it should be exactly same as from other files

File: fineract-core/src/main/java/org/apache/fineract/infrastructure/core/api/ApiParameterHelper.java [hunk 2]
```diff
@@ -58,15 +61,21 @@ public static Set<String> extractFieldsForResponseIfProvided(final MultivaluedMa
...
-            if (StringUtils.isNotBlank(commaSeparatedParameters)) {
-                fields = new HashSet<>(Arrays.asList(commaSeparatedParameters.split("\\s*,\\s*"))); // NOSONAR
+        Objects.requireNonNull(queryParams, "queryParams map cannot be null");
+        Set<String> associations = new HashSet<>();
+
+        // FIX: Change "fields" to "associations"
+        String commaSeparatedParameters = queryParams.getFirst("associations");
+
+        if (StringUtils.isNotBlank(commaSeparatedParameters)) {
+            String[] split = commaSeparatedParameters.split(",");
```
Comment (Aman-Mittal) [inline fineract-core/src/main/java/org/apache/fineract/infrastructure/core/api/ApiParameterHelper.java]: This will trigger errorProne String[] split = commaSeparatedParameters.split(",");  fix this

File: fineract-core/src/main/java/org/apache/fineract/infrastructure/core/api/ApiParameterHelper.java [hunk 3]
```diff
@@ -138,4 +147,24 @@ public static boolean genericResultSet(final MultivaluedMap<String, String> quer
     public static boolean genericResultSetPassed(final MultivaluedMap<String, String> queryParams) {
         return queryParams.getFirst(GENERIC_RESULT_SET) != null;
     }
+
+    public static Integer extractLimitParameter(MultivaluedMap<String, String> queryParameters) {
+        Objects.requireNonNull(queryParameters, "queryParameters map cannot be null");
+        String limit = queryParameters.getFirst("limit");
+        return (StringUtils.isNotBlank(limit)) ? Integer.valueOf(limit.trim()) : null;
```
Comment (Aman-Mittal) [inline fineract-core/src/main/java/org/apache/fineract/infrastructure/core/api/ApiParameterHelper.java]: /home/runner/work/fineract/fineract/fineract-validation/src/main/java/org/apache/fineract/validation/constraints/EnumValueValidator.java:33: warning: [StringCaseLocaleUsage] Specify a `Locale` when calling `String#to{Lower,Upper}Case`. (Note: there are multiple suggested fixes; the third may be most appropriate if you're dealing with ASCII Strings.)
        acceptedValues = Arrays.stream(annotation.enumClass().getEnumConstants()).map(e -> e.name().toLowerCase())
                                                                                                               ^
    (see https://errorprone.info/bugpattern/StringCaseLocaleUsage)
  Did you mean 'acceptedValues = Arrays.stream(annotation.enumClass().getEnumConstants()).map(e -> e.name().toLowerCase(Locale.ROOT))' or 'acceptedValues = Arrays.stream(annotation.enumClass().getEnumConstants()).map(e -> e.name().toLowerCase(Locale.getDefault()))' or 'acceptedValues = Arrays.stream(annotation.enumClass().getEnumConstants()).map(e -> Ascii.toLowerCase(e.name()))'?
/home/runner/work/fineract/fineract/fineract-validation/src/main/java/org/apache/fineract/validation/constraints/EnumValueValidator.java:39: warning: [StringCaseLocaleUsage] Specify a `Locale` when calling `String#to{Lower,Upper}Case`. (Note: there are multiple suggested fixes; the third may be most appropriate if you're dealing with ASCII Strings.)
        return value != null && acceptedValues.contains(value.toLowerCase());
                                                                         ^
    (see https://errorprone.info/bugpattern/StringCaseLocaleUsage)
  Did you mean 'return value != null && acceptedValues.contains(value.toLowerCase(Locale.ROOT));' or 'return value != null && acceptedValues.contains(value.toLowerCase(Locale.getDefault()));' or 'return value != null && acceptedValues.contains(Ascii.toLowerCase(value));'?
3 warnings
Unable to resolve POM for org.eclipse.platform:org.eclipse.swt:3.124.100

File: fineract-core/src/main/java/org/apache/fineract/infrastructure/core/api/ApiParameterHelper.java [hunk 4]
```diff
@@ -138,4 +147,24 @@ public static boolean genericResultSet(final MultivaluedMap<String, String> quer
...
+    }
+
+    public static Integer extractOffsetParameter(MultivaluedMap<String, String> queryParameters) {
+        Objects.requireNonNull(queryParameters, "queryParameters map cannot be null");
+        String offset = queryParameters.getFirst("offset");
```
Comment (Aman-Mittal) [inline fineract-core/src/main/java/org/apache/fineract/infrastructure/core/api/ApiParameterHelper.java]: @Ambika-Sony try running ./gradlew compileJava for errorProne for  more info.

Comment (Ambika-Sony) [reply fineract-core/src/main/java/org/apache/fineract/infrastructure/core/api/ApiParameterHelper.java]: > @Ambika-Sony try running ./gradlew compileJava for errorProne for more info.

Hi @Aman-Mittal, I've updated ApiParameterHelper.java to resolve the ErrorProne warnings.I replaced all split() regex calls with StringUtils.split() and manual trimming,refined the excludeAssociationsForResponseIfProvided logic to be more robust and Cleaned up imports and verified the build locally with ./gradlew compileJava.
Please let me know if any further changes are needed!

**Comment by Ambika-Sony** [comment]
Hi @Aman-Mittal, @adamsaghy — All checks are now passing (37/37). I have resolved the ErrorProne issues and verified the logic with the new unit tests. Ready for your final review when you have a moment!

File: fineract-core/src/main/java/org/apache/fineract/infrastructure/core/api/ApiParameterHelper.java [hunk 5]
```diff
@@ -46,38 +46,54 @@ public static Long commandId(final MultivaluedMap<String, String> queryParams) {
...
-        String commaSeparatedParameters = "";
-        if (queryParams.getFirst("associations") != null) {
-            commaSeparatedParameters = queryParams.getFirst("associations");
-            if (StringUtils.isNotBlank(commaSeparatedParameters)) {
-                fields = new HashSet<>(Arrays.asList(commaSeparatedParameters.split("\\s*,\\s*"))); // NOSONAR
+        Objects.requireNonNull(queryParams, "queryParams map cannot be null");
+        Set<String> associations = new HashSet<>();
+        String commaSeparatedParameters = queryParams.getFirst("associations");
+
+        if (StringUtils.isNotBlank(commaSeparatedParameters)) {
```
Comment (Aman-Mittal) [inline fineract-core/src/main/java/org/apache/fineract/infrastructure/core/api/ApiParameterHelper.java]: This seems duplication

File: fineract-core/src/test/java/org/apache/fineract/infrastructure/core/api/ApiParameterHelperTest.java [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+import jakarta.ws.rs.core.MultivaluedHashMap;
+import jakarta.ws.rs.core.MultivaluedMap;
+import java.util.Set;
+import org.junit.jupiter.api.Test;
+
+class ApiParameterHelperTest {
+
+    @Test
```
Comment (Aman-Mittal) [inline fineract-core/src/test/java/org/apache/fineract/infrastructure/core/api/ApiParameterHelperTest.java]: The current test coverage is a good start and covers the basic happy paths and invalid number parsing. However, I think we can strengthen this further by adding a few important edge cases to better protect API behavior.

Some scenarios that are currently not covered:

Multiple query parameters with the same key

e.g. ?fields=id,name&fields=description

Same for associations

Ensures proper flattening across multi-valued params.

Empty string values

?fields=

?associations=

Prevents accidental inclusion of empty entries.

Duplicate values

fields=id,name,id

Verifies deduplication behavior since the return type is Set.

Partial pagination parameters

Only offset present

Only limit present

Ensures default behavior is explicitly verified.

Negative pagination values

offset=-1, limit=-10

Clarifies whether negatives are allowed or should fail.

Overflow cases

Very large numeric values exceeding Integer range.

Important defensive test for parsing logic.

Adding these would make the test suite more robust and better document the intended contract of ApiParameterHelper, especially since these helpers are reused across many endpoints.

Comment (Ambika-Sony) [reply fineract-core/src/test/java/org/apache/fineract/infrastructure/core/api/ApiParameterHelperTest.java]: > The current test coverage is a good start and covers the basic happy paths and invalid number parsing. However, I think we can strengthen this further by adding a few important edge cases to better protect API behavior.
> 
> Some scenarios that are currently not covered:
> 
> Multiple query parameters with the same key
> 
> e.g. ?fields=id,name&fields=description
> 
> Same for associations
> 
> Ensures proper flattening across multi-valued params.
> 
> Empty string values
> 
> ?fields=
> 
> ?associations=
> 
> Prevents accidental inclusion of empty entries.
> 
> Duplicate values
> 
> fields=id,name,id
> 
> Verifies deduplication behavior since the return type is Set.
> 
> Partial pagination parameters
> 
> Only offset present
> 
> Only limit present
> 
> Ensures default behavior is explicitly verified.
> 
> Negative pagination values
> 
> offset=-1, limit=-10
> 
> Clarifies whether negatives are allowed or should fail.
> 
> Overflow cases
> 
> Very large numeric values exceeding Integer range.
> 
> Important defensive test for parsing logic.
> 
> Adding these would make the test suite more robust and better document the intended contract of ApiParameterHelper, especially since these helpers are reused across many endpoints.

Hi @Aman-Mittal,
I’ve updated the PR to address your feedback. The parameter extraction logic has been refactored to use a private helper to remove duplication, and it now properly handles multi-valued parameters. I’ve also expanded the test suite to cover deduplication, blank values, negative pagination, and overflow cases.

Please let me know if anything further should be adjusted.

**Comment by Ambika-Sony** [comment]
Hi, I've addressed all the edge cases and refactored the logic as
discussed. I noticed the workflows are currently awaiting approval to run.
Please let me know if you'd like me to adjust anything else once the builds
finish.

**Comment by adamsaghy** [comment]
@Ambika-Sony Please rebase this PR

**Comment by Ambika-Sony** [comment]
> @Ambika-Sony Please rebase this PR

I've rebased against the latest upstream and cleared out that conflict in the integration tests. I also made sure to keep the history clean with a single verified commit. Ready for another look.
